In [1]:
import pandas as pd
import numpy as np
from collections import OrderedDict
from gensim.parsing.preprocessing import STOPWORDS
from nltk.corpus import stopwords
import nltk
from nltk.stem import PorterStemmer

full_data = pd.read_csv('https://dso-560-nlp-text-analytics.s3.amazonaws.com/Full+data.csv')
extra_data = pd.read_csv('https://dso-560-nlp-text-analytics.s3.amazonaws.com/extra_data.csv')
sme_tag = pd.read_excel('https://dso-560-nlp-text-analytics.s3.amazonaws.com/USC+Product+Attribute+Data+03302020.xlsx')
additional_tag = pd.read_csv('https://dso-560-nlp-text-analytics.s3.amazonaws.com/usc_additional_tags.csv')
#attributes =  pd.read_excel('Womens+Attributes.xlsx')

/Users/dwko/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
full_data = full_data.iloc[:,:14]
full_data.head()

product_id                brand           mpn  \
0  01DSRPSZTDW2PGK1YWYXJGKZZ0                 FILA  400010319073   
1  01DSQXJBX0R7DCW7KTAC1SW547               CHANEL  400011497371   
2  01DPGV8TGRAB993PF7Z3YWG2VR                Frame      LWBT0099   
3  01DSR8G3F7DBRTMP8THF97XSQ2  Lilly Pulitzer Kids  400011844797   
4  01DSR8G5GP519DEDCSKBMWQVK5          Kissy Kissy  400094474585   

                                                name  \
0                          Original Fitness Sneakers   
1                                                HAT   
2                             Petit Oval Buckle Belt   
3  Little Gir's & Girl's Ariana One-Piece UPF 50+...   
4  Baby Girl's Endearing Elephants Pima Cotton Co...   

                                         description  \
0  Vintage Fitness leather sneakers with logo pri...   
1                                                NaN   
2  A Timeless Leather Belt Crafted From Smooth Co...   
3  Pretty ruffle sleeves and trim elevate essenti...   
4  Versatile convertible gown with elephant applique   

                                      brand_category  \
0                 TheMensStore/Shoes/Sneakers/LowTop   
1                                            Unknown   
2                                        Accessories   
3  JustKids/Girls214/Girls/SwimwearCoverups,JustK...   
4  JustKids/Baby024months/InfantGirls/FootiesRompers   

                      created_at                     updated_at  deleted_at  \
0   2019-11-15 23:36:38.98161+00  2019-12-19 20:40:30.786144+00         NaN   
1  2019-11-15 16:15:34.809725+00  2019-12-19 20:40:30.786144+00         NaN   
2  2019-10-06 15:33:38.451504+00  2019-12-19 20:40:30.786144+00         NaN   
3  2019-11-15 19:26:34.947426+00  2019-12-19 20:40:30.786144+00         NaN   
4  2019-11-15 19:26:36.995998+00  2019-12-19 20:40:30.786144+00         NaN   

                                 brand_canonical_url  \
0  https://www.saksfifthavenue.com/fila-original-...   
1  https://www.saksfifthavenue.com/chanel-hat/pro...   
2  https://frame-store.com/products/petit-oval-bu...   
3  https://www.saksfifthavenue.com/lilly-pulitzer...   
4  https://www.saksfifthavenue.com/kissy-kissy-ba...   

                                             details                labels  \
0  Leather/synthetic upper\nLace-up closure\nText...  {"Needs Attributes"}   
1                                  WOOL TWEED & FELT      {"Needs Review"}   
2                                                NaN                    {}   
3  Scoopneck\nAdjustable straps\nFlutter sleeves\...      {"Needs Review"}   
4  V-neckline\nLong sleeves\nFront snap closure\n...      {"Needs Review"}   

                                                 tsv  bc_product_id  
0  'design':12 'fila':1A 'fit':3A,6 'leather':7 '...            NaN  
1                               'chanel':1A 'hat':2A            NaN  
2  'belt':5A,9 'buckl':4A,21 'cowhid':13 'craft':...            NaN  
3  '50':14A 'allov':28 'ariana':9A 'color':27 'el...            NaN  
4  'appliqu':17 'babi':3A 'convert':10A,13 'cotto...            NaN

In [3]:
extra_data.tail()

product_id      mpn                       brand  \
6616  01E2KZ045BCT1Q2Z2PBC0CWRGH  5441202  ATM ANTHONY THOMAS MELILLO   
6617  01E2KXZ0WNQ96S64TDZE3WB7CX  5504588  ATM ANTHONY THOMAS MELILLO   
6618  01E2KXWWY6JAGENED7P62090NP  4496777  ATM ANTHONY THOMAS MELILLO   
6619  01E2KZTEPE7QK1V8THASF7HWNR  5289605  ATM ANTHONY THOMAS MELILLO   
6620  01E2KY2T6YV7R4VXT4BVV4TWT5  4788992  ATM ANTHONY THOMAS MELILLO   

     brand_category                               name labels  \
6616        Unknown          Micro Twill Pull On Pants     []   
6617        Unknown  Brushed Twill Crop Wide Leg Pants     []   
6618        Unknown                    Slim Crop Pants     []   
6619        Unknown              Camo Print Silk Skirt     []   
6620        Unknown       Stretch Pima Cotton Baby Tee     []   

                            updated_at  bc_product_id  \
6616  2020-03-26 00:06:42.342000+00:00           1957   
6617  2020-04-02 23:13:56.436000+00:00           1801   
6618  2020-04-02 23:11:55.640000+00:00           1791   
6619  2020-04-02 23:39:04.835000+00:00           2077   
6620  2020-04-02 23:33:25.438000+00:00           1818   

                              saleable  \
6616  2020-03-25 16:34:08.933000+00:00   
6617  2020-04-02 23:13:56.436000+00:00   
6618  2020-04-02 23:11:55.640000+00:00   
6619  2020-04-02 23:39:04.835000+00:00   
6620  2020-04-02 23:33:25.438000+00:00   

                                                details  \
6616  True to size. XS=0, S=2-4, M=6-8, L=10. High r...   
6617                           True to size. High rise.   
6618                            Mid rise. True to size.   
6619            True to size. XS=0, S=2-4, M=6-8, L=10.   
6620            True to size. XS=0, S=2-4, M=6-8, L=10.   

                            created_at  \
6616  2020-03-04 23:44:52.899000+00:00   
6617  2020-03-04 23:26:48.207000+00:00   
6618  2020-03-04 23:25:38.623000+00:00   
6619  2020-03-04 23:59:15.654000+00:00   
6620  2020-03-04 23:28:52.433000+00:00   

                                    brand_canonical_url notes  \
6616  https://shop.nordstrom.com/s/atm-anthony-thoma...   NaN   
6617  https://shop.nordstrom.com/s/atm-anthony-thoma...   NaN   
6618  https://shop.nordstrom.com/s/atm-anthony-thoma...   NaN   
6619  https://shop.nordstrom.com/s/atm-anthony-thoma...   NaN   
6620  https://shop.nordstrom.com/s/atm-anthony-thoma...   NaN   

                                            description  
6616  These casual trousers are cut for an easy, dra...  
6617  Brushed twill gives these cropped wide-leg pan...  
6618  Tailored from a soft, dense knit with cropped,...  
6619  Flatlock-stitched bias seams enhance the fluid...  
6620  Made from ATM’s signature Peruvian jersey, thi...

In [4]:
# Adding extra_data to full_data and removing the duplicates with same product_id

clean_df = pd.concat([full_data, extra_data], ignore_index=True)
clean_df.drop_duplicates(subset= ['product_id'], keep = 'first', inplace=True)
clean_df = clean_df[list(full_data.columns)]
clean_df.head()

/Users/dwko/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  This is separate from the ipykernel package so we can avoid doing imports until


product_id                brand           mpn  \
0  01DSRPSZTDW2PGK1YWYXJGKZZ0                 FILA  400010319073   
1  01DSQXJBX0R7DCW7KTAC1SW547               CHANEL  400011497371   
2  01DPGV8TGRAB993PF7Z3YWG2VR                Frame      LWBT0099   
3  01DSR8G3F7DBRTMP8THF97XSQ2  Lilly Pulitzer Kids  400011844797   
4  01DSR8G5GP519DEDCSKBMWQVK5          Kissy Kissy  400094474585   

                                                name  \
0                          Original Fitness Sneakers   
1                                                HAT   
2                             Petit Oval Buckle Belt   
3  Little Gir's & Girl's Ariana One-Piece UPF 50+...   
4  Baby Girl's Endearing Elephants Pima Cotton Co...   

                                         description  \
0  Vintage Fitness leather sneakers with logo pri...   
1                                                NaN   
2  A Timeless Leather Belt Crafted From Smooth Co...   
3  Pretty ruffle sleeves and trim elevate essenti...   
4  Versatile convertible gown with elephant applique   

                                      brand_category  \
0                 TheMensStore/Shoes/Sneakers/LowTop   
1                                            Unknown   
2                                        Accessories   
3  JustKids/Girls214/Girls/SwimwearCoverups,JustK...   
4  JustKids/Baby024months/InfantGirls/FootiesRompers   

                      created_at                     updated_at  deleted_at  \
0   2019-11-15 23:36:38.98161+00  2019-12-19 20:40:30.786144+00         NaN   
1  2019-11-15 16:15:34.809725+00  2019-12-19 20:40:30.786144+00         NaN   
2  2019-10-06 15:33:38.451504+00  2019-12-19 20:40:30.786144+00         NaN   
3  2019-11-15 19:26:34.947426+00  2019-12-19 20:40:30.786144+00         NaN   
4  2019-11-15 19:26:36.995998+00  2019-12-19 20:40:30.786144+00         NaN   

                                 brand_canonical_url  \
0  https://www.saksfifthavenue.com/fila-original-...   
1  https://www.saksfifthavenue.com/chanel-hat/pro...   
2  https://frame-store.com/products/petit-oval-bu...   
3  https://www.saksfifthavenue.com/lilly-pulitzer...   
4  https://www.saksfifthavenue.com/kissy-kissy-ba...   

                                             details                labels  \
0  Leather/synthetic upper\nLace-up closure\nText...  {"Needs Attributes"}   
1                                  WOOL TWEED & FELT      {"Needs Review"}   
2                                                NaN                    {}   
3  Scoopneck\nAdjustable straps\nFlutter sleeves\...      {"Needs Review"}   
4  V-neckline\nLong sleeves\nFront snap closure\n...      {"Needs Review"}   

                                                 tsv  bc_product_id  
0  'design':12 'fila':1A 'fit':3A,6 'leather':7 '...            NaN  
1                               'chanel':1A 'hat':2A            NaN  
2  'belt':5A,9 'buckl':4A,21 'cowhid':13 'craft':...            NaN  
3  '50':14A 'allov':28 'ariana':9A 'color':27 'el...            NaN  
4  'appliqu':17 'babi':3A 'convert':10A,13 'cotto...            NaN

In [5]:
len(clean_df)

48087

In [6]:
# I removed columns that would not be helpful for the text analytics (datetime)
train = clean_df.drop(columns=['created_at', 'updated_at', 'deleted_at', 'bc_product_id'])


# I made the entire dataframe in lowercase
for columns in train.columns:
    train[columns] = train[columns].str.lower() 
    
train.head()

product_id                brand           mpn  \
0  01dsrpsztdw2pgk1ywyxjgkzz0                 fila  400010319073   
1  01dsqxjbx0r7dcw7ktac1sw547               chanel  400011497371   
2  01dpgv8tgrab993pf7z3ywg2vr                frame      lwbt0099   
3  01dsr8g3f7dbrtmp8thf97xsq2  lilly pulitzer kids  400011844797   
4  01dsr8g5gp519dedcskbmwqvk5          kissy kissy  400094474585   

                                                name  \
0                          original fitness sneakers   
1                                                hat   
2                             petit oval buckle belt   
3  little gir's & girl's ariana one-piece upf 50+...   
4  baby girl's endearing elephants pima cotton co...   

                                         description  \
0  vintage fitness leather sneakers with logo pri...   
1                                                NaN   
2  a timeless leather belt crafted from smooth co...   
3  pretty ruffle sleeves and trim elevate essenti...   
4  versatile convertible gown with elephant applique   

                                      brand_category  \
0                 themensstore/shoes/sneakers/lowtop   
1                                            unknown   
2                                        accessories   
3  justkids/girls214/girls/swimwearcoverups,justk...   
4  justkids/baby024months/infantgirls/footiesrompers   

                                 brand_canonical_url  \
0  https://www.saksfifthavenue.com/fila-original-...   
1  https://www.saksfifthavenue.com/chanel-hat/pro...   
2  https://frame-store.com/products/petit-oval-bu...   
3  https://www.saksfifthavenue.com/lilly-pulitzer...   
4  https://www.saksfifthavenue.com/kissy-kissy-ba...   

                                             details                labels  \
0  leather/synthetic upper\nlace-up closure\ntext...  {"needs attributes"}   
1                                  wool tweed & felt      {"needs review"}   
2                                                NaN                    {}   
3  scoopneck\nadjustable straps\nflutter sleeves\...      {"needs review"}   
4  v-neckline\nlong sleeves\nfront snap closure\n...      {"needs review"}   

                                                 tsv  
0  'design':12 'fila':1a 'fit':3a,6 'leather':7 '...  
1                               'chanel':1a 'hat':2a  
2  'belt':5a,9 'buckl':4a,21 'cowhid':13 'craft':...  
3  '50':14a 'allov':28 'ariana':9a 'color':27 'el...  
4  'appliqu':17 'babi':3a 'convert':10a,13 'cotto...

In [7]:
# Combining the sme_tag and additional_tag
tag = pd.concat([sme_tag, additional_tag], ignore_index=True)
for columns in tag.columns:
    tag[columns] = tag[columns].str.lower() 

#the attribute name/vaue from sme_tag included punctuations including spaces, but additional_tag stored the data in one chunk of words
# ex) additonal color vs additionalcolor

# I decided to follow the additional_tag format by removing all punctuations and spaces for easier analysis moving on
remove = [' ', '_', '(', ')', '-', ',', '&', '"', '"', '/']
for i in remove:
    tag['attribute_name'] = tag['attribute_name'].str.replace(i, '')
    tag['attribute_value'] = tag['attribute_value'].str.replace(i, '')

tag.drop_duplicates(keep='first', inplace=True) #only removed duplicates that have same data in all of the columns
tag.reset_index(inplace = True)

In [8]:
tag['product_id'].nunique()

3970

In [9]:
# For the purpose of the project, we will only choose 4 out of these attribute names
attribute_name = list(tag.attribute_name.unique())
attribute_name.sort()
attribute_name

['additionalcolor',
 'beltbucklematerial',
 'beltbuckleshape',
 'beltclosure',
 'beltmaterial',
 'beltwidth',
 'calfwidth',
 'category',
 'classbelts',
 'classblazerscoatsandjackets',
 'classbooties',
 'classboots',
 'classdress',
 'classflats',
 'classhandbags',
 'classjumpsuitandromper',
 'classmulesandslides',
 'classpantsandleggings',
 'classpumpsandheels',
 'classsandals',
 'classshorts',
 'classskirts',
 'classslippers',
 'classsneakersandathletic',
 'classsunglasses',
 'classwedges',
 'closureblazerscoatsandjackets',
 'closurehandbag',
 'closureonepiece',
 'closurepantsandleggings',
 'closureshoe',
 'closureshorts',
 'closureskirts',
 'closuresweater',
 'closuretop',
 'color',
 'drycleanonly',
 'embellishment',
 'fit',
 'gender',
 'heelheight',
 'heelshape',
 'legstyle',
 'legstylejeans',
 'lengthblazers',
 'lengthblazerscoatsandjackets',
 'lengthcoatsandjackets',
 'lengthjeans',
 'lengthonepiece',
 'lengthpantsandleggings',
 'lengthshorts',
 'lengthskirts',
 'lengthtop',
 'mate

In [10]:
attribute_value = list(tag.attribute_value.unique())
attribute_value.sort()
attribute_value

['100%cashmere',
 '100%cotton',
 '100%linen',
 '100%silk',
 '100%wool',
 '14zip',
 '5pocketpantnondenim',
 'abstract',
 'accessory',
 'acetate',
 'acid',
 'acrylic',
 'active',
 'activetrack',
 'adjustable',
 'aline',
 'alpaca',
 'androgynous',
 'animal',
 'ankle',
 'anklestrap',
 'anorakraincoatwindbreaker',
 'asymmetrical',
 'athleisure',
 'aviatorsquare',
 'aviatorteardrop',
 'backless',
 'backpacks',
 'backtie',
 'backzip',
 'backzipper',
 'baggy',
 'ballet',
 'bandcollar',
 'beachbags',
 'beaded',
 'beiges',
 'belt',
 'beltbagsfannypack',
 'belted',
 'belts',
 'bermuda',
 'bike',
 'black',
 'blacks',
 'blazer',
 'blazerdress',
 'blazerscoatsjackets',
 'block',
 'blouse',
 'blues',
 'boatneck',
 'bodycon',
 'bodysuit',
 'boho',
 'bombervarsity',
 'bootcut',
 'bootie',
 'booties',
 'boots',
 'bottom',
 'boyfriend',
 'brass',
 'browns',
 'bucketbags',
 'buckle',
 'buckles',
 'burgundies',
 'businesscasual',
 'businesscasualdress',
 'bustier',
 'buttondown',
 'buttonedback',
 'buttone

# Out of all the attributes above, we are only going to use: 
- **style** 
- **occasion** 
- **category** 
- **fit**

In [11]:
# Creating new columns for the targeted categories
train['style'] = ""
train['occasion'] = ""
train['category'] = ""
train['fit'] = ""

In [12]:
group_tag = tag.groupby('product_id')
tagged_product = list(group_tag.groups.keys())

In [13]:
group_tag.get_group('01e2fdnd1xhvrj8p3whhkjgg1x')

index                  product_id            product_color_id  \
38797   45259  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
38798   45260  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
38799   45261  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
38800   45262  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
38801   45263  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95680  116601  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95681  116602  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95682  116603  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95683  116604  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95684  116605  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95685  116606  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95686  116607  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95687  116608  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95688  116609  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95689  116610  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95690  116611  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95691  116612  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95692  116613  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95693  116614  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   
95694  116615  01e2fdnd1xhvrj8p3whhkjgg1x  01e2fdnd27n1jekagccgmn7ngt   

        attribute_name attribute_value  
38797    uppermaterial           suede  
38798     primarycolor          blacks  
38799            style          casual  
38800            style            boho  
38801            style          modern  
95680       heelheight          medium  
95681         toestyle      pointedtoe  
95682         category            shoe  
95683      shaftheight         midcalf  
95684        heelshape            cone  
95685           gender           women  
95686       classboots   westerncowboy  
95687        shoewidth          medium  
95688      closureshoe            none  
95689         occasion      daytonight  
95690    embellishment          fringe  
95691         occasion         weekend  
95692  subcategoryshoe           boots  
95693       classboots          bootie  
95694        calfwidth          medium

#### This code will loop through each of the product_id from the tag.
#### Using the dataframe above, if the attribute_name includes any of the 4 focus attributes ('style', 'occasion', 'category', 'fit'), the code will add the corresponding attribute_value to the new columns created in the full_data + extra_data.

In [14]:
focus_attribute = ['style', 'occasion', 'category', 'fit']

for x in tagged_product:
    product = group_tag.get_group(x)
    focus = product.loc[product['attribute_name'].isin(focus_attribute)].reset_index()
    for i in range(0,len(focus)):
        if focus.loc[i, 'attribute_name'] == 'style':
            train.loc[train['product_id'] == x, 'style'] += (focus.loc[i, 'attribute_value'] + " ")
        elif focus.loc[i, 'attribute_name'] == 'occasion':
            train.loc[train['product_id'] == x, 'occasion'] += (focus.loc[i, 'attribute_value'] + " ")
        elif focus.loc[i, 'attribute_name'] == 'category':
            train.loc[train['product_id'] == x, 'category'] += (focus.loc[i, 'attribute_value'] + " ")
        elif focus.loc[i, 'attribute_name'] == 'fit':
            train.loc[train['product_id'] == x, 'fit'] += (focus.loc[i, 'attribute_value'] + " ")

In [15]:
# Removing duplicated words in a cell for the 4 columns
# source: https://stackoverflow.com/questions/47316783/python-dataframe-remove-duplicate-words-in-the-same-cell-within-a-column-in-pyt

for i in focus_attribute:
    train[i] = (train[i].str.split()
                              .apply(lambda x: OrderedDict.fromkeys(x).keys())
                              .str.join(' '))

The result for the 2 lines above can be observed here.

In [16]:
train

product_id                       brand           mpn  \
0      01dsrpsztdw2pgk1ywyxjgkzz0                        fila  400010319073   
1      01dsqxjbx0r7dcw7ktac1sw547                      chanel  400011497371   
2      01dpgv8tgrab993pf7z3ywg2vr                       frame      lwbt0099   
3      01dsr8g3f7dbrtmp8thf97xsq2         lilly pulitzer kids  400011844797   
4      01dsr8g5gp519dedcskbmwqvk5                 kissy kissy  400094474585   
...                           ...                         ...           ...   
48989  01e2kz045bct1q2z2pbc0cwrgh  atm anthony thomas melillo       5441202   
48990  01e2kxz0wnq96s64tdze3wb7cx  atm anthony thomas melillo       5504588   
48991  01e2kxwwy6jagened7p62090np  atm anthony thomas melillo       4496777   
48992  01e2kztepe7qk1v8thasf7hwnr  atm anthony thomas melillo       5289605   
48993  01e2ky2t6yv7r4vxt4bvv4twt5  atm anthony thomas melillo       4788992   

                                                    name  \
0                              original fitness sneakers   
1                                                    hat   
2                                 petit oval buckle belt   
3      little gir's & girl's ariana one-piece upf 50+...   
4      baby girl's endearing elephants pima cotton co...   
...                                                  ...   
48989                          micro twill pull on pants   
48990                  brushed twill crop wide leg pants   
48991                                    slim crop pants   
48992                              camo print silk skirt   
48993                       stretch pima cotton baby tee   

                                             description  \
0      vintage fitness leather sneakers with logo pri...   
1                                                    NaN   
2      a timeless leather belt crafted from smooth co...   
3      pretty ruffle sleeves and trim elevate essenti...   
4      versatile convertible gown with elephant applique   
...                                                  ...   
48989  these casual trousers are cut for an easy, dra...   
48990  brushed twill gives these cropped wide-leg pan...   
48991  tailored from a soft, dense knit with cropped,...   
48992  flatlock-stitched bias seams enhance the fluid...   
48993  made from atm’s signature peruvian jersey, thi...   

                                          brand_category  \
0                     themensstore/shoes/sneakers/lowtop   
1                                                unknown   
2                                            accessories   
3      justkids/girls214/girls/swimwearcoverups,justk...   
4      justkids/baby024months/infantgirls/footiesrompers   
...                                                  ...   
48989                                            unknown   
48990                                            unknown   
48991                                            unknown   
48992                                            unknown   
48993                                            unknown   

                                     brand_canonical_url  \
0      https://www.saksfifthavenue.com/fila-original-...   
1      https://www.saksfifthavenue.com/chanel-hat/pro...   
2      https://frame-store.com/products/petit-oval-bu...   
3      https://www.saksfifthavenue.com/lilly-pulitzer...   
4      https://www.saksfifthavenue.com/kissy-kissy-ba...   
...                                                  ...   
48989  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48990  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48991  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48992  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48993  https://shop.nordstrom.com/s/atm-anthony-thoma...   

                                                 details  \
0      leather/synthetic upper\nlace-up closure\ntext...   
1                                      wool tweed & felt   
2         

In [17]:
# Preparing to remove the stopwords

stopwords_gensim = list(STOPWORDS)

stopwords_NLTK = list(stopwords.words("english"))

stopwords_combined = list(set(stopwords_gensim+stopwords_NLTK)) #to remove duplicates

negatives = ['not','nor','no','neither', 'never'] #took out the negative words for a more accurate analysis
stopwords_combined = list(filter(lambda x: x not in negatives, stopwords_combined))

stopwords_combined.sort()
stopwords_expression = '|'.join(stopwords_combined)
stopwords_pattern = f'({stopwords_expression})'


In [18]:
# Defining function to stem the texts in each cell

def stem_text(text):
    porter=PorterStemmer()
    tokens = text.split()
    stemmed_tokens = [porter.stem(token) for token in tokens]
    return ' '.join(stemmed_tokens)



#source: https://stackoverflow.com/questions/43795310/apply-porters-stemmer-to-a-pandas-column-for-each-word

In [19]:
# For the columns shown below, this for loop will
# 1. remove all the punctuations and '\n' and only keep the words
# 2. remove all the stopwords defined above
# 3. perform stemming
# for each of the cells in the dataframe


# wasn't sure if we should remove brand in this list or not. Kept it for now.
col = ['brand', 'name', 'description', 'brand_category', 'details', 'labels', 'tsv']

for i in col:
    train[i] = train[i].astype(str)
    train[i] = train[i].str.replace('[^\w\s]',' ')
    train[i] = train[i].str.replace("\n", " ")
    train[i] = train[i].str.replace(rf'\b{stopwords_pattern}\b','')
    train[i] = train[i].apply(stem_text)

train

product_id                      brand           mpn  \
0      01dsrpsztdw2pgk1ywyxjgkzz0                       fila  400010319073   
1      01dsqxjbx0r7dcw7ktac1sw547                     chanel  400011497371   
2      01dpgv8tgrab993pf7z3ywg2vr                      frame      lwbt0099   
3      01dsr8g3f7dbrtmp8thf97xsq2           lilli pulitz kid  400011844797   
4      01dsr8g5gp519dedcskbmwqvk5                kissi kissi  400094474585   
...                           ...                        ...           ...   
48989  01e2kz045bct1q2z2pbc0cwrgh  atm anthoni thoma melillo       5441202   
48990  01e2kxz0wnq96s64tdze3wb7cx  atm anthoni thoma melillo       5504588   
48991  01e2kxwwy6jagened7p62090np  atm anthoni thoma melillo       4496777   
48992  01e2kztepe7qk1v8thasf7hwnr  atm anthoni thoma melillo       5289605   
48993  01e2ky2t6yv7r4vxt4bvv4twt5  atm anthoni thoma melillo       4788992   

                                                  name  \
0                                   origin fit sneaker   
1                                                  hat   
2                                petit oval buckl belt   
3           littl gir girl ariana piec upf 50 swimsuit   
4      babi girl endear eleph pima cotton convert gown   
...                                                ...   
48989                            micro twill pull pant   
48990                   brush twill crop wide leg pant   
48991                                   slim crop pant   
48992                            camo print silk skirt   
48993                     stretch pima cotton babi tee   

                                             description  \
0      vintag fit leather sneaker logo print design s...   
1                                                    nan   
2      timeless leather belt craft smooth cowhid mini...   
3      pretti ruffl sleev trim elev essenti tank swim...   
4                    versatil convert gown eleph appliqu   
...                                                  ...   
48989  casual trouser cut easi drapey fit thigh taper...   
48990    brush twill give crop wide leg pant subtl sheen   
48991  tailor soft dens knit crop cuf hem lean straig...   
48992  flatlock stitch bia seam enhanc fluid drape pu...   
48993  atm signatur peruvian jersey fit version label...   

                                          brand_category  \
0                        themensstor shoe sneaker lowtop   
1                                                unknown   
2                                              accessori   
3      justkid girls214 girl swimwearcoverup justkid ...   
4            justkid baby024month infantgirl footiesromp   
...                                                  ...   
48989                                            unknown   
48990                                            unknown   
48991                                            unknown   
48992                                            unknown   
48993                                            unknown   

                                     brand_canonical_url  \
0      https://www.saksfifthavenue.com/fila-original-...   
1      https://www.saksfifthavenue.com/chanel-hat/pro...   
2      https://frame-store.com/products/petit-oval-bu...   
3      https://www.saksfifthavenue.com/lilly-pulitzer...   
4      https://www.saksfifthavenue.com/kissy-kissy-ba...   
...                                                  ...   
48989  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48990  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48991  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48992  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48993  https://shop.nordstrom.com/s/atm-anthony-thoma...   

                                                 details         labels  \
0      leather synthet upper lace closur textil line ...  need attribut   
1                                        wool tweed felt    need review   
2

In [20]:
data = train
data = data.replace('nan',np.NaN)
data = data.replace('',np.NaN)
data_label = data.loc[(data['style'].notna()) | 
                  (data['occasion'].notna()) |
                  (data['category'].notna()) |
                  (data['fit'].notna())]
data_label.reset_index(inplace = True)

In [21]:
data

product_id                      brand           mpn  \
0      01dsrpsztdw2pgk1ywyxjgkzz0                       fila  400010319073   
1      01dsqxjbx0r7dcw7ktac1sw547                     chanel  400011497371   
2      01dpgv8tgrab993pf7z3ywg2vr                      frame      lwbt0099   
3      01dsr8g3f7dbrtmp8thf97xsq2           lilli pulitz kid  400011844797   
4      01dsr8g5gp519dedcskbmwqvk5                kissi kissi  400094474585   
...                           ...                        ...           ...   
48989  01e2kz045bct1q2z2pbc0cwrgh  atm anthoni thoma melillo       5441202   
48990  01e2kxz0wnq96s64tdze3wb7cx  atm anthoni thoma melillo       5504588   
48991  01e2kxwwy6jagened7p62090np  atm anthoni thoma melillo       4496777   
48992  01e2kztepe7qk1v8thasf7hwnr  atm anthoni thoma melillo       5289605   
48993  01e2ky2t6yv7r4vxt4bvv4twt5  atm anthoni thoma melillo       4788992   

                                                  name  \
0                                   origin fit sneaker   
1                                                  hat   
2                                petit oval buckl belt   
3           littl gir girl ariana piec upf 50 swimsuit   
4      babi girl endear eleph pima cotton convert gown   
...                                                ...   
48989                            micro twill pull pant   
48990                   brush twill crop wide leg pant   
48991                                   slim crop pant   
48992                            camo print silk skirt   
48993                     stretch pima cotton babi tee   

                                             description  \
0      vintag fit leather sneaker logo print design s...   
1                                                    NaN   
2      timeless leather belt craft smooth cowhid mini...   
3      pretti ruffl sleev trim elev essenti tank swim...   
4                    versatil convert gown eleph appliqu   
...                                                  ...   
48989  casual trouser cut easi drapey fit thigh taper...   
48990    brush twill give crop wide leg pant subtl sheen   
48991  tailor soft dens knit crop cuf hem lean straig...   
48992  flatlock stitch bia seam enhanc fluid drape pu...   
48993  atm signatur peruvian jersey fit version label...   

                                          brand_category  \
0                        themensstor shoe sneaker lowtop   
1                                                unknown   
2                                              accessori   
3      justkid girls214 girl swimwearcoverup justkid ...   
4            justkid baby024month infantgirl footiesromp   
...                                                  ...   
48989                                            unknown   
48990                                            unknown   
48991                                            unknown   
48992                                            unknown   
48993                                            unknown   

                                     brand_canonical_url  \
0      https://www.saksfifthavenue.com/fila-original-...   
1      https://www.saksfifthavenue.com/chanel-hat/pro...   
2      https://frame-store.com/products/petit-oval-bu...   
3      https://www.saksfifthavenue.com/lilly-pulitzer...   
4      https://www.saksfifthavenue.com/kissy-kissy-ba...   
...                                                  ...   
48989  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48990  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48991  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48992  https://shop.nordstrom.com/s/atm-anthony-thoma...   
48993  https://shop.nordstrom.com/s/atm-anthony-thoma...   

                                                 details         labels  \
0      leather synthet upper lace closur textil line ...  need attribut   
1                                        wool tweed felt    need review   
2

In [22]:
data_label

index                  product_id                      brand      mpn  \
0         6  01dtjcerf6f4nrz2wsjffa1eys                     theori  1197027   
1         8  01dvpbj6464ykygvae0a1hmkgn               alexand wang  1202011   
2        10  01dpetns8db4mar29er6esa7xa                     j crew    b5275   
3        11  01dt50zst2px9c1kkza4p73q5q               alexand wang  1154478   
4        17  01dvcqrrx4qyjv20667zk52s82                     theori  1205277   
...     ...                         ...                        ...      ...   
3965  48989  01e2kz045bct1q2z2pbc0cwrgh  atm anthoni thoma melillo  5441202   
3966  48990  01e2kxz0wnq96s64tdze3wb7cx  atm anthoni thoma melillo  5504588   
3967  48991  01e2kxwwy6jagened7p62090np  atm anthoni thoma melillo  4496777   
3968  48992  01e2kztepe7qk1v8thasf7hwnr  atm anthoni thoma melillo  5289605   
3969  48993  01e2ky2t6yv7r4vxt4bvv4twt5  atm anthoni thoma melillo  4788992   

                                                   name  \
0                             teah stretch silk camisol   
1                               layer velvet mini dress   
2                         faux leather pleat midi skirt   
3     parker logo embellish snake effect leather ank...   
4                                 silk crepe midi dress   
...                                                 ...   
3965                              micro twill pull pant   
3966                     brush twill crop wide leg pant   
3967                                     slim crop pant   
3968                              camo print silk skirt   
3969                       stretch pima cotton babi tee   

                                            description       brand_category  \
0     beig stretch silk slip 93 silk 7 spandex dri c...  cloth top tank cami   
1     black velvet conceal hook zip fasten 65 viscos...     cloth dress mini   
2     accordion pleat flatter midi length faux leath...                skirt   
3     heel measur approxim 75mm 3 inch snake effect ...       shoe boot ankl   
4     black silk crepe conceal zip fasten 100 silk l...     cloth dress midi   
...                                                 ...                  ...   
3965  casual trouser cut easi drapey fit thigh taper...              unknown   
3966    brush twill give crop wide leg pant subtl sheen              unknown   
3967  tailor soft dens knit crop cuf hem lean straig...              unknown   
3968  flatlock stitch bia seam enhanc fluid drape pu...              unknown   
3969  atm signatur peruvian jersey fit version label...              unknown   

                                    brand_canonical_url  \
0     https://www.net-a-porter.com/us/en/product/119...   
1     https://www.net-a-porter.com/us/en/product/120...   
2     https://www.jcrew.com/p/womens_category/skirts...   
3     https://www.net-a-porter.com/us/en/product/115...   
4     https://www.net-a-porter.com/us/en/product/120...   
...                                                 ...   
3965  https://shop.nordstrom.com/s/atm-anthony-thoma...   
3966  https://shop.nordstrom.com/s/atm-anthony-thoma...   
3967  https://shop.nordstrom.com/s/atm-anthony-thoma...   
3968  https://shop.nordstrom.com/s/atm-anthony-thoma...   
3969  https://shop.nordstrom.com/s/atm-anthony-thoma...   

                                                details         labels  \
0     fit true size normal size cut slightli loos fi...            NaN   
1     fit true size normal size design fit bust wais...            NaN   
2                                                   NaN            NaN   
3                fit true size normal size italian size  need attribut   
4     fit true size normal size cut slim fit drop wa...            NaN   
...                                                 ...            ...   
3965              true size xs 0 2 4 6 8 l 10 high rise            NaN   
3966                                true size high rise            NaN   
3967        

**I was worried about the limited size of the data with label (3,970 rows), which was less than 10% of the entire dataset (48,087 rows).**
However, when I checked the # of products that have been labeled through the sme_tag + additional_tag, there were only 3970 unique products, which matched our number of labelled products. This means that we were able to get all the labels for products that have been labelled by the SME.

In [23]:
tag['product_id'].nunique()

3970

In [24]:
data_label['product_id'].nunique()

3970